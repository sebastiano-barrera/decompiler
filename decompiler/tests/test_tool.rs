//! Tests based on "test_tool"
//!
//! These are pretty weak, but they work on real code. They only check that the
//! decompiler doesn't panic, and they write the decompiler's output as text to
//! a version-controlled file, just so that it's somewhere easy to diff and hard
//! to forget updating.

use include_dir::{include_dir, Dir};
use insta::assert_snapshot;

use std::{
    path::Path,
    sync::{Mutex, MutexGuard, OnceLock},
};

static DATA_DIR: Dir<'_> = include_dir!("$CARGO_MANIFEST_DIR/test-data/");

macro_rules! case {
    ($case:ident, $exe:expr, $func_name:ident) => {
        #[test_log::test]
        #[allow(non_snake_case)]
        fn $case() {
            let function_name = stringify!($func_name);
            let out = $exe.process_function(function_name);
            crate::assert_snapshot!(out);
        }
    };
}

macro_rules! tests_in_binary {
    ($group:ident, $exe_path:expr; $($funcs:ident),* $(,)?) => {
        mod $group {
            use crate::Exe;
            static EXE: Exe = Exe::new($exe_path);

            $(case!($funcs, EXE, $funcs);)*
        }
    };
}

struct Exe {
    once_lock: OnceLock<Mutex<decompiler::Executable<'static>>>,
    path: &'static str,
}
impl Exe {
    const fn new(path: &'static str) -> Self {
        Exe {
            once_lock: OnceLock::new(),
            path,
        }
    }

    fn get_or_init(&self) -> MutexGuard<'_, decompiler::Executable<'static>> {
        let mutex = self.once_lock.get_or_init(|| {
            let rel_path = Path::new(self.path);
            let raw = DATA_DIR.get_file(rel_path).unwrap().contents();
            let exe = decompiler::Executable::parse(raw).unwrap();
            Mutex::new(exe)
        });
        mutex.lock().unwrap()
    }

    fn process_function(&self, function_name: &str) -> String {
        use std::fmt::Write;

        let mut exe = self.get_or_init();

        let df = exe
            .decompile_function(function_name)
            .expect("decompiling function");

        let mut log_buf = String::new();

        if let Some(mil) = df.mil() {
            writeln!(log_buf, " --- mil").unwrap();
            writeln!(log_buf, "{:?}\n", mil).unwrap();
        }

        if let Some(ssa) = df.ssa_pre_xform() {
            writeln!(log_buf, " --- ssa pre-xform").unwrap();
            writeln!(log_buf, "{:?}\n", ssa).unwrap();
        }

        if let Some(ssa) = df.ssa() {
            writeln!(log_buf, " --- cfg").unwrap();
            let cfg = ssa.cfg();
            writeln!(log_buf, "  entry: {:?}", cfg.direct().entry_bid()).unwrap();
            for bid in cfg.block_ids() {
                let regs: Vec<_> = ssa.block_regs(bid).collect();
                writeln!(
                    log_buf,
                    "  {:?} -> {:?} {:?}",
                    bid,
                    cfg.block_cont(bid),
                    regs
                )
                .unwrap();
            }
            write!(log_buf, "  domtree:\n    ").unwrap();

            let mut pp_buf = decompiler::pp::FmtAsIoUTF8(&mut log_buf);
            let pp = &mut decompiler::pp::PrettyPrinter::start(&mut pp_buf);
            cfg.dom_tree().dump(pp).unwrap();
            writeln!(log_buf).unwrap();

            writeln!(log_buf, " --- ssa").unwrap();
            writeln!(log_buf, "{:?}\n", ssa).unwrap();

            writeln!(log_buf, " --- ast").unwrap();
            let mut ast = decompiler::Ast::new(&ssa, exe.types());
            let mut pp_buf = decompiler::pp::FmtAsIoUTF8(&mut log_buf);
            let pp = &mut decompiler::pp::PrettyPrinter::start(&mut pp_buf);
            ast.pretty_print(pp).unwrap();
        }

        log_buf
    }
}

tests_in_binary!(
    exe_composite_type, "ty/test_composite_type.so";
    list_len
);
tests_in_binary!(exe_capstone, "integration/sample_capstone"; main);
tests_in_binary!(exe_matrix, "integration/sample_matrix"; sum_matrix);
tests_in_binary!(
    redis_server, "integration/redis-server";
    ctl_lookup,
    listNext,
    listTypePush,
    geoArrayCreate,
    // regression: infinite loop
    processAnnotations,
    redisReconnect,
    redisConnectWithOptions,
);
