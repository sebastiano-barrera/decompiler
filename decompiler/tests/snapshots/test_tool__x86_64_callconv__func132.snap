---
source: decompiler/tests/test_tool.rs
expression: out
---
func132 0x4013bc+4 (file 0x13bc)
          4013bc: 44 89 c8                mov eax,r9d
          4013bf: c3                      ret

function type: "func132" func (arg0 *void <TypeID(1v1)>,
                               arg1 *void <TypeID(1v1)>,
                               arg2 *void <TypeID(1v1)>,
                               arg3 *void <TypeID(1v1)>,
                               arg4 *void <TypeID(1v1)>,
                               arg5 small <TypeID(784v1)>,
                               arg6 char <TypeID(778v1)>) char <TypeID(778v1)>
0 warnings.


mil program = 
program  92 instrs
                 0 r12 <- Ancestral(AncestralName("stack_bottom"))
                 1 r2 <- Ancestral(AncestralName("CF"))
                 2 r3 <- Ancestral(AncestralName("PF"))
                 3 r4 <- Ancestral(AncestralName("AF"))
                 4 r5 <- Ancestral(AncestralName("ZF"))
                 5 r6 <- Ancestral(AncestralName("SF"))
                 6 r7 <- Ancestral(AncestralName("TF"))
                 7 r8 <- Ancestral(AncestralName("IF"))
                 8 r9 <- Ancestral(AncestralName("DF"))
                 9 r10 <- Ancestral(AncestralName("OF"))
                10 r11 <- Ancestral(AncestralName("RBP"))
                11 r12 <- Ancestral(AncestralName("RSP"))
                12 r13 <- Ancestral(AncestralName("RIP"))
                13 r14 <- Ancestral(AncestralName("RDI"))
                14 r15 <- Ancestral(AncestralName("RSI"))
                15 r16 <- Ancestral(AncestralName("RAX"))
                16 r17 <- Ancestral(AncestralName("RBX"))
                17 r18 <- Ancestral(AncestralName("RCX"))
                18 r19 <- Ancestral(AncestralName("RDX"))
                19 r20 <- Ancestral(AncestralName("R8"))
                20 r21 <- Ancestral(AncestralName("R9"))
                21 r22 <- Ancestral(AncestralName("R10"))
                22 r23 <- Ancestral(AncestralName("R11"))
                23 r24 <- Ancestral(AncestralName("R12"))
                24 r25 <- Ancestral(AncestralName("R13"))
                25 r26 <- Ancestral(AncestralName("R14"))
                26 r27 <- Ancestral(AncestralName("R15"))
                27 r28 <- Ancestral(AncestralName("ZMM0"))
                28 r29 <- Ancestral(AncestralName("ZMM1"))
                29 r30 <- Ancestral(AncestralName("ZMM2"))
                30 r31 <- Ancestral(AncestralName("ZMM3"))
                31 r32 <- Ancestral(AncestralName("ZMM4"))
                32 r33 <- Ancestral(AncestralName("ZMM5"))
                33 r34 <- Ancestral(AncestralName("ZMM6"))
                34 r35 <- Ancestral(AncestralName("ZMM7"))
                35 r36 <- Ancestral(AncestralName("ZMM8"))
                36 r37 <- Ancestral(AncestralName("ZMM9"))
                37 r38 <- Ancestral(AncestralName("ZMM10"))
                38 r39 <- Ancestral(AncestralName("ZMM11"))
                39 r40 <- Ancestral(AncestralName("ZMM12"))
                40 r41 <- Ancestral(AncestralName("ZMM13"))
                41 r42 <- Ancestral(AncestralName("ZMM14"))
                42 r43 <- Ancestral(AncestralName("ZMM15"))
                43 r45 <- Undefined
                44 r46 <- Undefined
                45 r47 <- Undefined
                46 r48 <- Undefined
                47 r49 <- Undefined
                48 r50 <- Undefined
                49 r51 <- Undefined
                50 r52 <- Undefined
                51 r53 <- Undefined
                52 r54 <- Undefined
                53 r55 <- Undefined
                54 r56 <- Undefined
                55 r57 <- Undefined
                56 r58 <- Undefined
                57 r59 <- Undefined
                58 r60 <- Undefined
                59 r61 <- Undefined
                60 r62 <- Undefined
                61 r63 <- Undefined
                62 r64 <- Undefined
                63 r65 <- Undefined
                64 r45 <- Ancestral(AncestralName("arg0"))
                65 r14 <- Get(r45)
                66 r46 <- Ancestral(AncestralName("arg1"))
                67 r15 <- Get(r46)
                68 r47 <- Ancestral(AncestralName("arg2"))
                69 r19 <- Get(r47)
                70 r48 <- Ancestral(AncestralName("arg3"))
                71 r18 <- Get(r48)
                72 r49 <- Ancestral(AncestralName("arg4"))
                73 r20 <- Get(r49)
                74 r50 <- Ancestral(AncestralName("arg5"))
                75 r52 <- Part { src: r50, offset: 0, size: 8 }
                76 r51 <- ArithK(Add, r12, 8)
                77 r51 <- StoreMem { addr: r51, value: r52 }
                78 r52 <- Part { src: r50, offset: 8, size: 8 }
                79 r51 <- ArithK(Add, r12, 16)
                80 r51 <- StoreMem { addr: r51, value: r52 }
                81 r53 <- Ancestral(AncestralName("arg6"))
                82 r54 <- Widen { reg: r53, target_size: 8, sign: false }
                83 r21 <- Get(r54)
0x4013bc:
                84 r46 <- Part { src: r21, offset: 0, size: 4 }
                85 r47 <- Part { src: r16, offset: 4, size: 4 }
                86 r16 <- Concat { lo: r46, hi: r47 }
0x4013bf:
                87 r45 <- Void
                88 r45 <- Concat { lo: r45, hi: r16 }
                89 r45 <- Part { src: r45, offset: 0, size: 1 }
                90 r47 <- SetReturnValue(r45)
                91 r47 <- Control(Ret)


ssa pre-xform:
ssa program  92 instrs
.B0:    ;;  → Always(Return).
  (  2)  r11 <- Ancestral(AncestralName("RSP"))
         r15 <- Ancestral(AncestralName("RAX"))
  (  2)  r74 <- Ancestral(AncestralName("arg5"))
         r75 <- Part { src: r74, offset: 0, size: 8 }
         r76 <- ArithK(Add, r11, 8)
         r77 <- StoreMem { addr: r76, value: r75 }
         r78 <- Part { src: r74, offset: 8, size: 8 }
         r79 <- ArithK(Add, r11, 16)
         r80 <- StoreMem { addr: r79, value: r78 }
         r81 <- Ancestral(AncestralName("arg6"))
         r82 <- Widen { reg: r81, target_size: 8, sign: false }
         r84 <- Part { src: r82, offset: 0, size: 4 }
         r85 <- Part { src: r15, offset: 4, size: 4 }
         r86 <- Concat { lo: r84, hi: r85 }
         r88 <- Concat { lo: r87, hi: r86 }
         r89 <- Part { src: r88, offset: 0, size: 1 }
         r90 <- SetReturnValue(r89)


cfg:
  entry: BlockID(0)
  BlockID(0) -> Always(Return) [r11, r15, r74, r75, r76, r77, r78, r79, r80, r81, r82, r84, r85, r86, r87, r88, r89, r90]
  domtree:
    BlockID(0)

ssa post-xform:
ssa program  92 instrs
.B0:    ;;  → Always(Return).
  (  2)  r11 <- Ancestral(AncestralName("RSP"))
  (  2)  r74 <- Ancestral(AncestralName("arg5"))
         r75 <- Part { src: r74, offset: 0, size: 8 }
         r76 <- ArithK(Add, r11, 8)
         r77 <- StoreMem { addr: r76, value: r75 }
         r78 <- Part { src: r74, offset: 8, size: 8 }
         r79 <- ArithK(Add, r11, 16)
         r80 <- StoreMem { addr: r79, value: r78 }
         r81 <- Ancestral(AncestralName("arg6"))
         r90 <- SetReturnValue(r81)



T0: {
  [pre:RSP + 8]:* := pre:arg5[8 .. 0];
  [pre:RSP + 16]:* := pre:arg5[16 .. 8];
  ;
  return pre:arg6
}
